### PR TITLE
Use NetAnalyzers from SDK instead of FxCopAnalyzers

### LINKED ISSUE #30099: Migrate from FxCopAnalyzers to NetAnalyzers
FxCopAnalyzers package is deprecated. The repository should use `Microsoft.CodeAnalysis.NetAnalyzers` instead.

See [Migration guide](https://docs.microsoft.com/visualstudio/code-quality/migrate-from-fxcop-analyzers-to-net-analyzers).

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #30099

- [x] ~~Replaces FxCopAnalyzers package references with NetAnalyzers.~~ (Used from .NET SDK instead of package)
- [x] Deletes rulset files
- [x] Migrate rulesets to editorconfigs

@sharwell Can you review?

### CODE REVIEW

File: eng/targets/CSharp.Common.props
Comment (Youssef1313) [inline eng/targets/CSharp.Common.props:16]: I'm not sure what is this reference supposed to do when it's already in `Dependencies.props`

**Comment by Youssef1313** [comment]
There CA1310 violations like:

https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentBindLoweringPass.cs#L499

Do you want me to pass a StringComparison.Ordinal, or StringComparison.Invariant, or disable the rule?

*I'm not sure why the `main` branch doesn't fail with this error. It's already enabled without my changes :smile:

File: src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/ClientApp/.editorconfig
Comment (Pilchie) [inline src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/ClientApp/.editorconfig:2]: @javiercn - is this actually part of the template content, and should be left alone?

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/SourceSpan.cs
Comment (Pilchie) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/SourceSpan.cs:88]: Note - safe because `SourceSpan` is a struct, so can't be null, and will never be `ReferenceEquals`.

**Review by Pilchie — COMMENTED** [review]
@pranavkm or @NTaylorMullen any ideas on why it doesn't already fail?

I definitely prefer `Ordinal` comparisons whenever possible, especially for language keyword comparisons.

**Comment by Pilchie** [comment]
Looks good so far - let me know when you're done.

File: src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/ClientApp/.editorconfig
Comment (javiercn) [reply src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/ClientApp/.editorconfig:2]: Yes, we should leave this alone

File: .editorconfig
Comment (dougbu) [inline .editorconfig:42]: Can any (better many) of these work without `:suggestion`❔ I realize a few rules e.g. `csharp_style_var_elsewhere` may need to be easily overridden but hope that's not the case for most.

Comment (Pilchie) [reply .editorconfig:42]: Are you proposing something stronger (i.e. warning), or something weaker (i.e. no indication in the IDE at all)?

Comment (dougbu) [reply .editorconfig:42]: `warning` or `true` (I don't remember the distinction offhand)

Comment (Pilchie) [reply .editorconfig:42]: `true/false` is whether to prefer these, or prefer the opposite.  The thing after the `:` is the enforcement level:

`none, silent, suggestion, warning, error`.

Comment (dougbu) [reply .editorconfig:42]: So simple😃 My preference is `true:warning` and `false:warning` wherever possible i.e. unless it causes lots 'o churn in our codebase.

Comment (Pilchie) [reply .editorconfig:42]: Since we have `<TreatWarningsAsErrors>true</>` that ends up being an error anyway, which is pretty strong. It means your build fails and you can't even run tests if you have that wrong. Maybe we'll get there, but I have a feeling people might not like it if we go straight to errors in one go.

Comment (dougbu) [reply .editorconfig:42]: That makes sense. At some point when I'm not busy, I might try changing a few of these to see if everything builds w/o making changes in the sources.

File: .editorconfig
Comment (sharwell) [inline .editorconfig:74]: 💡 Setting diagnostic severity is more efficient if it's done in a globalconfig instead of editorconfig. See https://github.com/dotnet/aspnetcore-tooling/pull/3087 for an example of how I converted aspnetcore-tooling from rulesets to globalconfig.

**Comment by Youssef1313** [comment]
Due to large number of violations for many rules, I'm switching from AllEnabledByDefault to Default mode. Waiting on CI to see how many violations will be, fix them, review deleted rulesets, and convert them to global configs as @sharwell suggested.

Global configs are indeed much better as they integrate very well with MSBuild properties (for example, they allow having different rules easily for tests vs. product code).

File: .editorconfig
Comment (Youssef1313) [reply .editorconfig:42]: @Pilchie Warnings as errors should be in CI only.

If it's not already in CI only, I'd suggest doing so. This will give developers more flexibility running tests while warnings are there, and will also ensure that no warnings goes into the `main` branch.

File: src/Mvc/MvcTests.ruleset
Comment (Youssef1313) [inline src/Mvc/MvcTests.ruleset:7]: There are some violations of this rule. The rule should be enabled and violations should get fixed as they make tests very confusing.

For example, the "expected" parameter of a test isn't used.

However, this is out of scope for this PR.

Do you want me to open a tracking  issue?

File: src/Servers/IIS/AspNetCoreModuleV2/DefaultRules.ruleset
Comment (Youssef1313) [inline src/Servers/IIS/AspNetCoreModuleV2/DefaultRules.ruleset:11]: How was I seeing these as "CAxxxx"? 😄 

@sharwell These rules are for C++. Should I keep this ruleset as-is?

File: src/Mvc/Directory.Build.props
Comment (Youssef1313) [inline src/Mvc/Directory.Build.props:5]: @sharwell I'm still getting warnings from a disabled analyzer. Can you give a look?

**Comment by Youssef1313** [comment]
There are some violations of CA1305 from <https://github.com/aspnet/MessagePack-CSharp.git> which is imported through gitmodules.

File: src/Mvc/MvcTests.ruleset
Comment (dougbu) [reply src/Mvc/MvcTests.ruleset:7]: > Do you want me to open a tracking issue?

No, this isn't an issue we can actually fix. Some test parameters are used for their types and not their values. Fortunately, there should be very few violations.

File: .editorconfig
Comment (dougbu) [reply .editorconfig:42]: > Warnings as errors should be in CI only.

This repo is not configured that way. For one thing, we don't want the CI to surprise users with warnings they didn't see locally.

File: src/Mvc/MvcTests.ruleset
Comment (Youssef1313) [reply src/Mvc/MvcTests.ruleset:7]: @dougbu Can you elaborate more?

Here:

https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Mvc/Mvc.Formatters.Xml/test/SerializableErrorWrapperProviderTest.cs#L11-L21

`isSerialization` parameter is never used. The test might be wrong if it should use it, or it should be removed.

**Comment by Youssef1313** [comment]
This should be ready for another look.

File: .editorconfig
Comment (Pilchie) [reply .editorconfig:42]: I've used it both ways, and typically prefer the WarnAsError on CI, but I don't think this PR is the place to change it for the repo.  Something we can talk about at @dotnet/aspdoi maybe and get more opinions on.

File: src/Testing/src/xunit/SkipOnHelixAttribute.cs
Comment (Pilchie) [inline src/Testing/src/xunit/SkipOnHelixAttribute.cs:55]: @HaoK Should this be `Ordinal`, or `OrdianalIgnoreCase` do you think?

**Review by Pilchie — APPROVED** [review]
This looks good to me, but I'd like @pranavkm or @NTaylorMullen to make sure they agree with me about the `Ordinal` comparisons in the Razor compiler before we merge.

File: src/Testing/src/xunit/SkipOnHelixAttribute.cs
Comment (HaoK) [reply src/Testing/src/xunit/SkipOnHelixAttribute.cs:55]: OrdinalIgnoreCase seems more appropriate, since we don't actually care about casing differences

Comment (Youssef1313) [reply src/Testing/src/xunit/SkipOnHelixAttribute.cs:55]:
```suggestion
            if (Queues.Contains("All.OSX") && targetQueue.StartsWith("osx", StringComparison.OrdinalIgnoreCase))
```

File: src/Mvc/MvcTests.ruleset
Comment (dougbu) [reply src/Mvc/MvcTests.ruleset:7]: I didn't say nothing could be fixed😁

File: src/Testing/src/xunit/SkipOnHelixAttribute.cs
Comment (pranavkm) [inline src/Testing/src/xunit/SkipOnHelixAttribute.cs:55]:
```suggestion
            if (Queues.Contains("All.OSX") && targetQueue.StartsWith("osx", StringComparison.Ordinal))
```

That would be the equivalent to the previous check

**Review by pranavkm — APPROVED** [review]
Thanks, this is absolutely great!

File: src/Testing/src/xunit/SkipOnHelixAttribute.cs
Comment (pranavkm) [reply src/Testing/src/xunit/SkipOnHelixAttribute.cs:55]: I'll let this be. This is test-only code so it doesn't make a difference.

Comment (Pilchie) [reply src/Testing/src/xunit/SkipOnHelixAttribute.cs:55]: @HaoK actually asked to change this to OrdinalIgnoreCase as part of this review.

**Pull request merged**
